compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -236,137 +142,640 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        // Otherwise, if an argument is unsatisfiable, indicate that it's missing
+                        (_, true, _, true) => return Some(Issue::Missing(i)),
+                        (true, true, _, _) => {
+                            // The argument isn't useless, and the input isn't unsatisfied,
+                            // so look for a parameter we might swap it with
+                            // We look for swaps explicitly, instead of just falling back on permutations
+                            // so that cases like (A,B,C,D) given (B,A,D,C) show up as two swaps,
+                            // instead of a large permutation of 4 elements.
+                            for j in 0..cmp::min(ai.len(), ii.len()) {
+                                if i == j || mat[j][j] {
+                                    continue;
+                                }
+                                if mat[i][j] && mat[j][i] {
+                                    return Some(Issue::Swap(i, j));
+                                }
+                            }
+                        }
+                        _ => {
+                            continue;
+                        }
+                    };
                 }
 
-                debug!("checking the argument");
-                let formal_ty = formal_tys[i];
+                // We didn't find any of the individual issues above, but
+                // there might be a larger permutation of parameters, so we now check for that
+                // by checking for cycles
+                // We use a double option at position i in this vec to represent:
+                // - None: We haven't computed anything about this argument yet
+                // - Some(None): This argument definitely doesn't participate in a cycle
+                // - Some(Some(x)): the i-th argument could permute to the x-th position
+                let mut permutation: Vec<Option<Option<usize>>> = vec![None; mat.len()];
+                let mut permutation_found = false;
+                for i in 0..mat.len() {
+                    if permutation[i].is_some() {
+                        // We've already decided whether this argument is or is not in a loop
+                        continue;
+                    }
 
-                // The special-cased logic below has three functions:
-                // 1. Provide as good of an expected type as possible.
-                let expected = Expectation::rvalue_hint(self, expected_arg_tys[i]);
+                    let mut stack = vec![];
+                    let mut j = i;
+                    let mut last = i;
+                    let mut is_cycle = true;
+                    loop {
+                        stack.push(j);
+                        // Look for params this one could slot into
+                        let compat: Vec<_> =
+                            mat[j].iter().enumerate().filter(|(_, &c)| c).map(|(i, _)| i).collect();
```

Comment:
```suggestion
                            mat[j].iter().enumerate().filter_map(|(i, &c)| if c { Some(i) } else { None }).collect();
```
(or https://doc.rust-lang.org/std/primitive.bool.html#method.then_some if you want to be fancy)